pkg/river/parser/interp_string.go

```diff
@@ -0,0 +1,282 @@
+// left of it.
+//
+// interpStringScanner supports the same string escape sequences as normal
+// strings.
+type interpStringScanner struct {
+	file       *token.File
+	input      []byte
+	err        scanner.ErrorHandler // Error reporting (may be nil)
+	baseOffset int
+
+	ch         rune // Current character
+	offset     int  // Byte offset of ch
+	readOffset int  // Byte offset of first character *after* ch
+	numErrors  int  // Number of errors encountered during scanning.
+}
+
+// interpStringToken is a token for interpolated strings.
+type interpStringToken int
+
+const (
+	interpStringTokenIllegal interpStringToken = iota
+	interpStringTokenEOF
+
+	interpStringTokenRaw  // Hello, world!
+	interpStringTokenExpr // ${INTERPOLATED_EXPR}
+)
+
+const (
+	bom = 0xFEFF // byte order mark, permitted as very first character
+	eof = -1     // end of file
+)
+
+// newInterpStringScanner creates a new interpolated string scanner to tokenize
+// the provided input config.
+//
+// input must be a subset of text in file containing an interpolated string.
+// baseOff must be the offset of input within the original source for file.
+// baseOff will be added to all returned position offsets from Scan.
+//
+// Calls to Scan will invoke the error handler eh when a lexical error is found
+// if eh is not nil.
+func newInterpStringScanner(file *token.File, input []byte, eh scanner.ErrorHandler, baseOff int) *interpStringScanner {
+	s := &interpStringScanner{
+		file:       file,
+		input:      input,
+		err:        eh,
+		baseOffset: baseOff,
+	}
+
+	// Preload first character. Note that BOM are not handled since input never
+	// corresponds to a physical file.
+	s.next()
+	return s
+}
+
+// peek gets the next byte after the current character without advancing the
+// scanner. Returns 0 if the scanner is at EOF.
+func (s *interpStringScanner) peek() byte {
+	if s.readOffset < len(s.input) {
+		return s.input[s.readOffset]
+	}
+	return 0
+}
+
+// next advances the scanner and reads the next Unicode character into s.ch.
+// s.ch == eof indicates end of file.
+func (s *interpStringScanner) next() {
+	if s.readOffset >= len(s.input) {
+		s.offset = len(s.input)
+		if s.ch == '\n' {
+			// Make sure we track final newlines at the end of the file
+			s.file.AddLine(s.offset)
+		}
+		s.ch = eof
+		return
+	}
+
+	s.offset = s.readOffset
+	if s.ch == '\n' {
+		s.file.AddLine(s.offset)
+	}
+
+	r, width := rune(s.input[s.readOffset]), 1
+	switch {
+	case r == 0:
+		s.onError(s.offset, "illegal character NUL")
+	case r >= utf8.RuneSelf:
+		r, width = utf8.DecodeRune(s.input[s.readOffset:])
+		if r == utf8.RuneError && width == 1 {
+			s.onError(s.offset, "illegal UTF-8 encoding")
+		} else if r == bom && s.offset > 0 {
+			s.onError(s.offset, "illegal byte order mark")
+		}
+	}
+	s.readOffset += width
+	s.ch = r
+}
+
+func (s *interpStringScanner) onError(offset int, msg string) {
+	if s.err != nil {
+		s.err(s.file.Pos(offset), msg)
+	}
+	s.numErrors++
+}
```

Comment:
These utility methods are all copied verbatim from river/scanner, but I don't think it makes sense to try to refactor them out into a shared location.